src/control/server/server_utils.go

```diff
@@ -196,34 +196,81 @@ func updateFabricEnvars(log logging.Logger, cfg *engine.Config, fis *hardware.Fa
 	// Mercury will now support the new OFI_DOMAIN environment variable so
 	// that we can specify the correct device for each.
 	if !cfg.HasEnvVar("OFI_DOMAIN") {
-		fi, err := fis.GetInterfaceOnNetDevice(cfg.Fabric.Interface, cfg.Fabric.Provider)
+		interfaces, err := cfg.Fabric.GetInterfaces()
 		if err != nil {
-			return errors.Wrapf(err, "unable to determine device domain for %s", cfg.Fabric.Interface)
+			return err
+		}
+
+		providers, err := cfg.Fabric.GetProviders()
+		if err != nil {
+			return err
+		}
+
+		if len(providers) != len(interfaces) {
+			return errors.New("number of providers not equal to number of interfaces")
+		}
+
+		domains := []string{}
+
+		for i, p := range providers {
+			fi, err := fis.GetInterfaceOnNetDevice(interfaces[i], p)
+			if err != nil {
+				return errors.Wrapf(err, "unable to determine device domain for %s", interfaces[i])
+			}
+			domains = append(domains, fi.Name)
 		}
-		log.Debugf("setting OFI_DOMAIN=%s for %s", fi.Name, cfg.Fabric.Interface)
-		envVar := "OFI_DOMAIN=" + fi.Name
+
+		domain := strings.Join(domains, engine.MultiProviderSeparator)
+		log.Debugf("setting OFI_DOMAIN=%s for %s", domain, cfg.Fabric.Interface)
+		envVar := "OFI_DOMAIN=" + domain
 		cfg.WithEnvVars(envVar)
 	}
 
 	return nil
 }
 
-func getFabricNetDevClass(cfg *config.Server, fis *hardware.FabricInterfaceSet) (hardware.NetDevClass, error) {
-	var netDevClass hardware.NetDevClass
+func getFabricNetDevClass(cfg *config.Server, fis *hardware.FabricInterfaceSet) ([]hardware.NetDevClass, error) {
+	netDevClass := []hardware.NetDevClass{}
 	for index, engine := range cfg.Engines {
-		fi, err := fis.GetInterfaceOnNetDevice(engine.Fabric.Interface, engine.Fabric.Provider)
+		cfgIfaces, err := engine.Fabric.GetInterfaces()
 		if err != nil {
-			return 0, err
+			return nil, err
 		}
 
-		ndc := fi.DeviceClass
-		if index == 0 {
-			netDevClass = ndc
-			continue
+		provs, err := engine.Fabric.GetProviders()
+		if err != nil {
+			return nil, err
+		}
+
+		if len(provs) == 1 {
+			for i := range cfgIfaces {
+				if i == 0 {
+					continue
+				}
+				provs = append(provs, provs[0])
+			}
+		}
+
+		if len(cfgIfaces) != len(provs) {
+			return nil, fmt.Errorf("number of ifaces (%d) and providers (%d) not equal",
+				len(cfgIfaces), len(provs))
 		}
-		if ndc != netDevClass {
-			return 0, config.FaultConfigInvalidNetDevClass(index, netDevClass,
-				ndc, engine.Fabric.Interface)
+
+		for i, cfgIface := range cfgIfaces {
+			fi, err := fis.GetInterfaceOnNetDevice(cfgIface, provs[i])
```

Comment:
We should cover this in the unit test cases by adding a multi-provider element in the tests.